### PR TITLE
tar: Complete adding files to archive with long options

### PR DESCRIPTION
**Pull request by yedayak**
Fixes: https://github.com/scop/bash-completion/issues/98

### CODE REVIEW

**Review by akinomyoga — COMMENTED** [review]
Thanks for the fixes! I see `FIXME` comments, but are you going to work on these within this PR?

**Comment by yedayak** [comment]
Added a commit that removes the `FIXMEs` by keeping track of compression long options

File: completions/tar [hunk 1]
```diff
@@ -437,10 +444,18 @@ _comp_cmd_tar__cleanup_prev()
     fi
 }
 
+_comp_cmd_tar__is_bsdtar()
```
Comment (yedayak) [inline completions/tar:447]: I'm not sure about this, does this make sense as a function?
If not, I'm not sure where to define it as a variable

Comment (scop) [reply completions/tar:447]: I think it's fine as a function. Its contents could be reduced to just this though:
```
[[ ${COMP_WORDS[0]} == ?(*/)bsdtar ]]
```

File: completions/tar [hunk 2]
```diff
@@ -453,33 +468,33 @@ _comp_cmd_tar__detect_ext()
...
+        : # Pass through using defaults above
+    elif [[ $tar_mode_arg == *[Zz]*f || $compression_mode == gzip || $compression_mode == compress ]]; then
+        ext='@(@(t?(ar.)|spkg.)@(gz|Z)|taz)'
+    elif [[ $tar_mode_arg == *[jy]*f || $compression_mode == bzip2 ]]; then
+        ext='@(t?(ar.)bz?(2)|spkg|tb2)'
+    elif [[ $tar_mode_arg == *[J]*f || $compression_mode == xz ]]; then
+        ext='@(@(tar|spkg).@(lzma|xz)|t[lx]z)'
+    elif [[ $compression_mode == zstd ]]; then
+        ext='t?(ar.)zst'
+    fi
```
Comment (scop) [inline completions/tar:497]: These could be more readable when turned into `case $tar_mode_arg in` blocks.

Comment (yedayak) [reply completions/tar:494]: Do you mean putting `if [[ $compression_mode ==` inside each case? Or something else?

Comment (akinomyoga) [reply completions/tar:494]: You could do `case $tar_mode_arg:$compression_mode in` and match it to `+([^ZzJjy])f:none` or `*[Zz]*f:* | *:gzip | *:compress`, etc.

Comment (scop) [reply completions/tar:494]: Yep, that's even better :+1:

Comment (yedayak) [reply completions/tar]: Did that, it conflicted a bit with the bsdtar change, so I put it in the same commit with the compression change.

**Review by akinomyoga — CHANGES_REQUESTED** [review]

File: completions/tar [hunk 3]
```diff
@@ -453,33 +465,47 @@ _comp_cmd_tar__detect_ext()
...
-        *[jy]*f)
-            ext='@(t?(ar.)bz?(2)|spkg|tb2)'
-            ;;
-        *[J]*f)
-            ext='@(@(tar|spkg).@(lzma|xz)|t[lx]z)'
-            ;;
-    esac
+        fi
+        case "$tar_mode_arg:$compression_mode" in
+            "*a*:none" | "*:auto-compress")
```
[the remark this answers is not in the document]
Comment (akinomyoga) [reply completions/tar]: Thanks for updating!

File: completions/tar [hunk 4]
```diff
@@ -453,33 +465,47 @@ _comp_cmd_tar__detect_ext()
...
-            ;;
-        *[jy]*f)
-            ext='@(t?(ar.)bz?(2)|spkg|tb2)'
-            ;;
-        *[J]*f)
-            ext='@(@(tar|spkg).@(lzma|xz)|t[lx]z)'
-            ;;
-    esac
+        fi
+        case "$tar_mode_arg:$compression_mode" in
```
[the remark this answers is not in the document]
Comment (akinomyoga) [reply completions/tar]: Thanks.

File: completions/tar [hunk 5]
```diff
@@ -455,33 +465,47 @@ _comp_cmd_tar__detect_ext()
...
+            *J*:none | *:xz)
+                ext='t?(ar.)xz'
+                ;;
+        esac
+    else
+        #TODO: lzip, lzma, lzop
+        case "$tar_mode_arg:$compression_mode" in
+            +([^ZzJjy])f:none)
+                :
+                ;; # Pass through using defaults above
```
[the remark this answers is not in the document]
Comment (yedayak) [reply completions/tar]: removed

File: completions/tar [hunk 6]
```diff
@@ -453,33 +465,47 @@ _comp_cmd_tar__detect_ext()
...
+            *[Zz]*f:none | *:gzip | *:compress)
+                ext='@(@(t?(ar.)|spkg.)@(gz|Z)|taz)'
+                ;;
+            *[jy]*f:none | *:bzip2)
+                ext='@(t?(ar.)bz?(2)|spkg|tb2)'
+                ;;
+            *[J]*f:none | *:xz)
```
[the remark this answers is not in the document]
Comment (akinomyoga) [reply completions/tar]: Thanks.

File: completions/tar (hunk 5 above)
[the remark this answers is not in the document]
Comment (akinomyoga) [reply completions/tar]: Thank you!

File: completions/tar (hunk 3 above)
[the remark this answers is not in the document]
Comment (yedayak) [reply completions/tar]: Yeah started with quotes, changed, and forgot to remove these. Fixed

File: completions/tar (hunk 4 above)
[the remark this answers is not in the document]
Comment (yedayak) [reply completions/tar]: Fixed

File: completions/tar (hunk 6 above)
[the remark this answers is not in the document]
Comment (yedayak) [reply completions/tar]: Done

File: completions/tar [hunk 7]
```diff
@@ -494,9 +517,11 @@ _comp_cmd_tar__gnu()
     local tar_mode=none
 
     # The mode argument, e.g. -cpf or -c
-    # FIXME: handle long options
     local tar_mode_arg=
 
+    # Compression mode - from long options
+    local compression_mode=none
```
[the remark this answers is not in the document]
Comment (yedayak) [reply completions/tar]: Makes sense, changed to `tar_compression_mode`

[the remark this answers is not in the document]
Comment (akinomyoga) [reply completions/tar]: Thank you.

**Review by akinomyoga — APPROVED** [review]

**Comment by yedayak** [comment]
Rebased, removed a useless assignment of `tar_mode`, and improved the `tar_mode` comment

**Comment by yedayak** [comment]
Is there something left to do here?

**Review by akinomyoga — APPROVED** [review]

**Pull request merged**